Do not automatically enable 'developer' features during tests

brentfulgham opened the pull request:
#### 1a660cb880b4f36210f34d76055019f5e2057896
<pre>
Do not automatically enable &apos;developer&apos; features during tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=250235">https://bugs.webkit.org/show_bug.cgi?id=250235</a>
&lt;rdar://problem/103969233&gt;

Reviewed by Elliott Williams.

The generator logic added in Bug 247926 also activated features with &apos;developer&apos;
status while running tests.

This is wrong for a couple of reasons:
(1) The &apos;developer&apos; flags with a defaultValue of &apos;false&apos; are generally not relevant to our
automated testing (e.g., enabling security-violating CORS state, enabling features not meant
for web use cases, etc.
(2) TestWebKitAPI tests that need these special settings already enable the individually, so
we should not hard-set them to &apos;true&apos; at each test iteration, as this can also enable features
on platforms or configurations that need these features to be off.

This patch limits the &quot;automatically enabled&quot; feature flags to those that are marked as
&apos;preview&apos;, &apos;testable&apos;, or &apos;stable&apos;.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.cpp.erb:

Canonical link: <a href="https://commits.webkit.org/258681@main">https://commits.webkit.org/258681@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3bcedbaa3fd41e5fd90422625eb0e940cb49e75b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102646 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111913 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172136 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12781 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2682 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94921 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109614 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108422 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9808 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37466 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/106172 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91660 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24537 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79209 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92880 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5232 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25956 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89207 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2907 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5392 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2402 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29636 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11413 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45445 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92131 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5964 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7122 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20638 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple reviewed (approved):
Nice and simple. Looks good, though we should ensure this passes tests (i.e. we aren't depending on a developer/preview setting being turned on).

nt1m commented on Source/WTF/Scripts/GeneratePreferences.rb:
```diff
@@ -208,6 +208,11 @@ def experimental?
+  
+  # Features which should be enabled while running tests
+  def testable?
+    %w{ testable preview }.include? @status
```
`stable` should be treated  like `preview`. There's not much difference between the two. 

(Also this is what is causing the GTK failures)

brentfulgham replied:
No -- stable should be kept in its default state. If GTK relied on something in the ExperimentalFeatures yaml file to be turned on for testing, it should use 'testable' state.

nt1m replied:
I think this is problematic because you can't have per-platform statuses. Hence I think maybe the best solution is to add `stable` to that list. Either way, it doesn't make much sense to me to treat `preview` differently from `stable`, either both should be kept, or removed. `preview` is just another stable configuration.

brentfulgham replied:
No -- `preview` indicates a feature that is not yet on by default in shipping Safari, but should be on in STP. This is a very important distinction.

For now, it seems like enabling `stable` features is this simplest action at present.

Likely Elliott and I will need to add a port-specific feature for state so we can have distinct settings for each port.

Commit:
Do not automatically enable 'developer' features during tests
https://bugs.webkit.org/show_bug.cgi?id=250235
<rdar://problem/103969233>

Reviewed by Elliott Williams.

The generator logic added in Bug 247926 also activated features with 'developer'
status while running tests.

This is wrong for a couple of reasons:
(1) The 'developer' flags with a defaultValue of 'false' are generally not relevant to our
automated testing (e.g., enabling security-violating CORS state, enabling features not meant
for web use cases, etc.
(2) TestWebKitAPI tests that need these special settings already enable the individually, so
we should not hard-set them to 'true' at each test iteration, as this can also enable features
on platforms or configurations that need these features to be off.

This patch limits the "automatically enabled" feature flags to those that are marked as
'preview', 'testable', or 'stable'.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.cpp.erb:

Canonical link: https://commits.webkit.org/258681@main
